code/modules/cm_aliens/hivebuffs/hivebuff.dm

```diff
@@ -0,0 +1,517 @@
+			pylons_to_use += potential_pylon
+
+	if(!_check_can_afford_buff())
+		to_chat(purchasing_mob, SPAN_XENONOTICE("Our hive cannot afford [name]! [hive.buff_points] / [cost] points."))
+		return FALSE
+
+	if(!_check_pass_active())
+		to_chat(purchasing_mob, SPAN_XENONOTICE("[name] is already active in our hive!"))
+		return FALSE
+
+	if(!_check_pass_reusable())
+		to_chat(purchasing_mob, SPAN_XENONOTICE("Our hive has already used [name] and cannot use it again!"))
+		return FALSE
+
+	if(!_check_pass_unique())
+		var/active_buffs = ""
+		for(var/buff in hive.active_hivebuffs)
+			active_buffs += buff + " "
+		active_buffs = trim_right(active_buffs)
+		to_chat(purchasing_mob, SPAN_XENONOTICE("[name] cannot be used with other active buffs! Wait for those to end first. Active buffs: [active_buffs]"))
+		return FALSE
+
+	if(!handle_special_checks())
+		to_chat(purchasing_mob, SPAN_XENONOTICE(special_fail_message))
+		return FALSE
+
+	log_admin("[purchasing_mob] of [hive.hivenumber] is attempting to purchase a hive buff: [name].")
+
+	if(!_seek_queen_approval(purchasing_mob))
+		to_chat(purchasing_mob, SPAN_XENONOTICE("Our queen has not approved the purchase of [name]."))
+		return FALSE
+
+	// _seek_queen_approval() includes a 20 second timeout so we check that everything still exists that we need.
+	if(QDELETED(purchased_pylon) || QDELETED(purchasing_mob) && !purchasing_mob.check_state())
+		return FALSE
+
+	// Actually process the buff and apply effects - If the buff succeeds engage_message will return TRUE, if it fails there should be an engage_failure_message set.
+	if(!on_engage(purchased_pylon))
+		if(engage_failure_message && istext(engage_failure_message))
+			to_chat(purchasing_mob, SPAN_XENONOTICE(engage_failure_message))
+			to_chat(hive.living_xeno_queen, SPAN_XENONOTICE(engage_failure_message))
+			return
+		else
+			stack_trace("[purchasing_mob] attempted to purchase a hive buff: [name] and failed to engage and returned an invalid failure message or no failure message.")
+			return
+
+	// All checks have passed.
+
+	// Purchase and deduct funds only after we're sure the buff has engaged
+	if(!_purchase_and_deduct(purchasing_mob))
+		return
+
+	for(var/mob/living/carbon/xenomorph/xeno in hive.totalXenos)
+		if(apply_on_new_xeno)
+			RegisterSignal(SSdcs, COMSIG_GLOB_XENO_SPAWN, PROC_REF(_handle_xenomorph_new))
+		apply_buff_effects(xeno)
+
+
+	log_admin("[purchasing_mob] and [hive.living_xeno_queen] of [hive.hivenumber] have purchased a hive buff: [name].")
+
+	// Add to the relevant hive lists.
+	LAZYADD(hive.used_hivebuffs, src)
+	LAZYADD(hive.active_hivebuffs, src)
+
+	// Register signal to check if the pylon is ever destroyed.
+
+	// for(var/obj/effect/alien/resin/special/pylon/endgame/pylon_to_register in pylons_to_use)
+	// 	LAZYADD(sustained_pylons, purchased_pylon)
+	// 	pylon_to_register.sustain_hivebuff(src)
+	// 	RegisterSignal(pylon_to_register, COMSIG_PARENT_QDELETING, PROC_REF(_on_pylon_deletion))
```

Comment:
Why did we opt to check count of pylons in a timer rather than as signal here?

Reply:
Because the destroyer buff has no duration, it just spawns the destroyer and disappears